### PR TITLE
ARROW-9654: [Rust][DataFusion] Add `EXPLAIN <SQL>` statement

### PR DESCRIPTION
**Pull request by alamb**
In order to help users and developers understand what DataFusion's planner is doing, this PR adds an `"EXPLAIN PLAN"` feature. All other database systems I have worked with have such a feature (e.g. see [MySql](https://dev.mysql.com/doc/refman/8.0/en/explain-output.html)). 

Example printout (the plans printed are simply the `std::fmt::Debug` representation of the plan structures:)

```
> explain SELECT status, COUNT(1) FROM http_api_requests_total WHERE path = '/api/v2/write' GROUP BY status;
+--------------+----------------------------------------------------------------------+
| plan_type    | plan                                                                 |
+--------------+----------------------------------------------------------------------+
| logical_plan | Aggregate: groupBy=[[#status]], aggr=[[COUNT(UInt8(1))]]             |
|              |   Selection: #path Eq Utf8("/api/v2/write") And #path Eq Utf8("foo") |
|              |     TableScan: http_api_requests_total projection=None               |
+--------------+----------------------------------------------------------------------+
1 rows in set. Query took 0 seconds.
```
and

```
> explain VERBOSE SELECT status, COUNT(1) FROM http_api_requests_total WHERE path = '/api/v2/write' GROUP BY status;

+-----------------------------------------+----------------------------------------------------------------------+
| plan_type                               | plan                                                                 |
+-----------------------------------------+----------------------------------------------------------------------+
| logical_plan                            | Aggregate: groupBy=[[#status]], aggr=[[COUNT(UInt8(1))]]             |
|                                         |   Selection: #path Eq Utf8("/api/v2/write") And #path Eq Utf8("foo") |
|                                         |     TableScan: http_api_requests_total projection=None               |
| logical_plan after projection_push_down | Aggregate: groupBy=[[#status]], aggr=[[COUNT(UInt8(1))]]             |
|                                         |   Selection: #path Eq Utf8("/api/v2/write") And #path Eq Utf8("foo") |
|                                         |     TableScan: http_api_requests_total projection=Some([6, 8])       |
| logical_plan after type_coercion        | Aggregate: groupBy=[[#status]], aggr=[[COUNT(UInt8(1))]]             |
|                                         |   Selection: #path Eq Utf8("/api/v2/write") And #path Eq Utf8("foo") |
|                                         |     TableScan: http_api_requests_total projection=Some([6, 8])       |
| physical_plan                           | HashAggregateExec {                                                  |
|                                         |     group_expr: [                                                    |
|                                         |         Column {                                                     |
|                                         |             name: "status",                                          |
|                                         |         },                                                           |
|                                         |     ],                                                               |
|                                         |     aggr_expr: [                                                     |
|                                         |         Count {                                                      |
|                                         |             expr: Literal {                                          |
|                                         |                 value: UInt8(                                        |
|                                         |                     1,                                               |
|                                         |                 ),                                                   |
|                                         |             },                                                       |
|                                         |         },                                                           |
|                                         |     ],                                                               |
|                                         |     input: SelectionExec {                                           |
|                                         |         expr: BinaryExpr {                                           |
|                                         |             left: BinaryExpr {                                       |
|                                         |                 left: Column {                                       |
|                                         |                     name: "path",                                    |
|                                         |                 },                                                   |
|                                         |                 op: Eq,                                              |
|                                         |                 right: Literal {                                     |
|                                         |                     value: Utf8(                                     |
|                                         |                         "/api/v2/write",                             |
|                                         |                     ),                                               |
|                                         |                 },                                                   |
|                                         |             },                                                       |
|                                         |             op: And,                                                 |
|                                         |             right: BinaryExpr {                                      |
|                                         |                 left: Column {                                       |
|                                         |                     name: "path",                                    |
|                                         |                 },                                                   |
|                                         |                 op: Eq,                                              |
|                                         |                 right: Literal {                                     |
|                                         |                     value: Utf8(                                     |
|                                         |                         "foo",                                       |
|                                         |                     ),                                               |
|                                         |                 },                                                   |
|                                         |             },                                                       |
|                                         |         },                                                           |
|                                         |         input: DataSourceExec {                                      |
|                                         |             schema: Schema {                                         |
|                                         |                 fields: [                                            |
|                                         |                     Field {                                          |
|                                         |                         name: "path",                                |
|                                         |                         data_type: Utf8,                             |
|                                         |                         nullable: true,                              |
|                                         |                         dict_id: 0,                                  |
|                                         |                         dict_is_ordered: false,                      |
|                                         |                     },                                               |
|                                         |                     Field {                                          |
|                                         |                         name: "status",                              |
|                                         |                         data_type: Utf8,                             |
|                                         |                         nullable: true,                              |
|                                         |                         dict_id: 0,                                  |
|                                         |                         dict_is_ordered: false,                      |
|                                         |                     },                                               |
|                                         |                 ],                                                   |
|                                         |                 metadata: {},                                        |
|                                         |             },                                                       |
|                                         |             partitions.len: 1,                                       |
|                                         |         },                                                           |
|                                         |     },                                                               |
|                                         |     schema: Schema {                                                 |
|                                         |         fields: [                                                    |
|                                         |             Field {                                                  |
|                                         |                 name: "status",                                      |
|                                         |                 data_type: Utf8,                                     |
|                                         |                 nullable: true,                                      |
|                                         |                 dict_id: 0,                                          |
|                                         |                 dict_is_ordered: false,                              |
|                                         |             },                                                       |
|                                         |             Field {                                                  |
|                                         |                 name: "COUNT(UInt8(1))",                             |
|                                         |                 data_type: UInt64,                                   |
|                                         |                 nullable: true,                                      |
|                                         |                 dict_id: 0,                                          |
|                                         |                 dict_is_ordered: false,                              |
|                                         |             },                                                       |
|                                         |         ],                                                           |
|                                         |         metadata: {},                                                |
|                                         |     },                                                               |
|                                         | }                                                                    |
+-----------------------------------------+----------------------------------------------------------------------+


4 row in set. Query took 0 seconds.

```

### CODE REVIEW

File: rust/datafusion/src/logicalplan.rs
Comment (alamb) [inline rust/datafusion/src/logicalplan.rs:789]: In a subsequent PR, I would like to suggest we consider making this `Explain(ExplainNode)` or something, and having a separate `struct ExplainNode` with these fields. There is a bunch of boiler plate code for matching and reconstructing `LogicalPlanner` nodes that I think would be less awkward (and require fewer changes to add / modify fields)  if they were collected into a struct

File: rust/datafusion/src/optimizer/projection_push_down.rs
Comment (alamb) [inline rust/datafusion/src/optimizer/projection_push_down.rs:53]: I needed to make this `mut` so that `optimize_explain` can call `OptimizerRule::optimize` which requires a `mut` reference

**Comment by alamb** [comment]
I am sure this PR conflicts with https://github.com/apache/arrow/pull/7880... FYI @jorgecarleitao

**Comment by andygrove** [comment]
This makes me very happy! I took a quick skim through and LGTM but will review more tomorrow.

File: rust/datafusion/tests/sql.rs
Comment (jorgecarleitao) [inline rust/datafusion/tests/sql.rs:444]: `with lowercase explain`?

File: rust/datafusion/tests/sql.rs
Comment (jorgecarleitao) [inline rust/datafusion/tests/sql.rs:456]: yes! :)

**Review by jorgecarleitao — COMMENTED** [review]
I reviewed this PR.

Looks really good and it is definitely useful! Thanks for it, @alamb !

I left minor comments, and I have 4 questions:

#### Physical Plan?

We are declaring `EXPLAIN` as a LogicalPlan and then convert it to a Physical. I always understood EXPLAIN as a different type of statement, like the CreateExternalTable, the main reason being that it does not depend on data, just on the plan it explains, options, and the optimizers to considered.

Looking at CreateExternalTable, we do not construct a physical plan out of it. Wouldn't it be simpler if we do not create a physical plan, like we do for CreateExternalTable, since we do not need to multi-thread it nor run through partitions nor batches?

#### Plans as {:#?}

If I read correctly, we currently return the result of the explain as the debug (`{:#?}`) of the plans. Shouldn't we reserve the debug format for our internal representation, and introduce a new format to `EXPLAIN` (not necessarily in this PR)?

#### Other formats?

One thing I dislike about spark is that EXPLAIN is in a not-so-nice format to parse programmatically. One idea is to allow `explain` to return the result in json. This is e.g. useful to construct a visual graph representation of the plan. One idea would be to support another string parameter, `format`, and make it default to `string` (not necessarily in this PR).

#### Support on `table` API?

We should consider adding `.explain` on the `table.rs`.

**Comment by alamb** [comment]
Thanks @jorgecarleitao 

> #### Physical Plan?
> Looking at CreateExternalTable, we do not construct a physical plan out of it. Wouldn't it be simpler if we do not create a physical plan, like we do for CreateExternalTable, since we do not need to multi-thread it nor run through partitions nor batches?

Yes, you are right that we could special case explain not to actually run a physical plan. I was thinking there are some benefits to using a physical plan, as follows:

1. Immediate support for all other input/output channels (not just the CLI) -- so for example if there is ever a client library for DataFusion, it will not need any additional logic to handle `EXPLAIN`
2. Use / store the plans themselves as data ( imagine `INSERT INTO foo EXPLAIN select * from bar`)/  

I admit these are both forward looking features with likely no practical import in the near term. However, I felt making `Explain` a first class part of both logical and physical plans would reduce special cases and make the code "cleaner" (though this is just my opinion)

> #### Plans as {:#?}
> If I read correctly, we currently return the result of the explain as the debug (`{:#?}`) of the plans. Shouldn't we reserve the debug format for our internal representation, and introduce a new format to `EXPLAIN` (not necessarily in this PR)?

I am torn on this and have no strong opinion. I could go either way -- I think that most of the information that is typically in explain plans is useful for developers, and thus there would always be substantial overlap between a Debug format and an explain format. But maybe the differences between them could justify the duplication? I am not sure. 

> #### Other formats?
> One thing I dislike about spark is that EXPLAIN is in a not-so-nice format to parse programmatically. One idea is to allow `explain` to return the result in json. This is e.g. useful to construct a visual graph representation of the plan. One idea would be to support another string parameter, `format`, and make it default to `string` (not necessarily in this PR).

I like this idea. The typical thing I have used in the past for such purposes is https://graphviz.org/ and `dot` -- I have filed https://issues.apache.org/jira/browse/ARROW-9746 to track the improvement. 

> #### Support on `table` API?
> We should consider adding `.explain` on the `table.rs`.

Good idea -- I defer to @andygrove on how he sees the DataFusion CLI evolving -- aka if he wants to mirror the SQLite style `.table` or `.quit` type commands or the current `quit` command or something else entirely.

File: rust/datafusion/tests/sql.rs
Comment (alamb) [reply rust/datafusion/tests/sql.rs:444]: Yes, good catch. Will fix

File: rust/datafusion/src/logicalplan.rs
Comment (alamb) [reply rust/datafusion/src/logicalplan.rs:1177]: I think an enum might be cleaner too -- let me give it a shit

File: rust/datafusion/tests/sql.rs
Comment (alamb) [reply rust/datafusion/tests/sql.rs]: fixed in https://github.com/apache/arrow/pull/7959/commits/639db5f109c5861c7fb646152a48b156afde4ac3

File: rust/datafusion/src/logicalplan.rs
Comment (alamb) [reply rust/datafusion/src/logicalplan.rs:1203]: @jorgecarleitao  -- check this out: https://github.com/apache/arrow/pull/7959/commits/7e9e42ce385a4825ff986b70f87960cec6a18f1c

**Comment by jorgecarleitao** [comment]
I agree with your answers, the changes, and ready to merge for me.

It is being great fun working on this string of PRs from each of us, @alamb and @andygrove ! 👍

**Comment by alamb** [comment]
> It is being great fun working on this string of PRs from each of us, @alamb and @andygrove ! 👍

I agree @jorgecarleitao

**Review by andygrove — APPROVED** [review]
Thanks @alamb and @jorgecarleitao this looks great. I will have some things to suggest as follow-ups most likely but nothing that should block this from being merged.

**Pull request closed without merging**
